internal/controller/controllers/preprovisioningimage_controller.go

```diff
@@ -400,25 +403,29 @@ func (r *PreprovisioningImageReconciler) AddIronicAgentToInfraEnv(ctx context.Co
 	conf, err := ignition.GenerateIronicConfig(r.IronicServiceURL, r.IronicInspectorURL, *infraEnvInternal, ironicAgentImage)
 	if err != nil {
 		log.WithError(err).Error("failed to generate Ironic ignition config")
-		return ctrl.Result{}, err
+		return false, err
 	}
 
-	_, err = r.Installer.UpdateInfraEnvInternal(ctx, installer.UpdateInfraEnvParams{InfraEnvID: *infraEnvInternal.ID, InfraEnvUpdateParams: &models.InfraEnvUpdateParams{}}, swag.String(string(conf)))
-	if err != nil {
-		return ctrl.Result{}, err
-	}
-	if infraEnv.ObjectMeta.Annotations == nil {
-		infraEnv.ObjectMeta.Annotations = make(map[string]string)
-	}
+	updated := false
+	if string(conf) != infraEnvInternal.InternalIgnitionConfigOverride {
+		_, err = r.Installer.UpdateInfraEnvInternal(ctx, installer.UpdateInfraEnvParams{InfraEnvID: *infraEnvInternal.ID, InfraEnvUpdateParams: &models.InfraEnvUpdateParams{}}, swag.String(string(conf)))
+		if err != nil {
+			return false, err
+		}
+		updated = true
 
-	infraEnv.Annotations[EnableIronicAgentAnnotation] = "true"
-	err = r.Client.Update(ctx, infraEnv)
-	if err != nil {
-		return ctrl.Result{}, err
+		if infraEnv.ObjectMeta.Annotations == nil {
+			infraEnv.ObjectMeta.Annotations = make(map[string]string)
+		}
+		infraEnv.Annotations[EnableIronicAgentAnnotation] = "true"
+		if err := r.Client.Update(ctx, infraEnv); err != nil {
+			// Just warn here if the update fails as this annotation is just informational
+			log.WithError(err).Warnf("failed to set %s annotation on infraEnv %s", EnableIronicAgentAnnotation, infraEnv.Name)
+		}
+
+		r.CRDEventsHandler.NotifyInfraEnvUpdates(infraEnv.Name, infraEnv.Namespace)
```

Comment:
doesn't adding annotation trigger reconcile?

Reply:
Maybe but for one, we may already have this annotation so the update may be a noop. This was also there before, just the indentation changed.

I can leave the comment too, if you want :man_shrugging:

Reply:
The code is confusing because there is not check that the annotation is actually set or not, it may already exist. 
So the notification is actually depend on the `updated` field